Add cheat for easy pause buffering

garrettjoecox opened the pull request:
This is to help players achieve those precise frame perfect tricks, you can find the setting under Cheats > Easy Pause Buffering. This is done in a bit of an unconventional way in that all input is disabled and then re-enabled on the first frame after pause. What this means is there is no time window you need to wait for to start holding your inputs, you can start holding them before unpausing if desired. 
The one special case here is the START button, there is not much use in buffering a START input as it would simply re-pause the game on the same frame, so START input is buffered one extra frame, so if you simply continue holding START when unpausing it will advance one frame, and re-pause the game.

https://user-images.githubusercontent.com/7316699/189051227-38f66586-5bae-4b4a-ba80-29a5000aefe0.mov

garrettjoecox committed:
Add cheat for easy pause buffering

garrettjoecox commented:
After doing some more research, I think a better term for this cheat is easy frame advancing, but I'm open to input

PurpleHato commented:
Pause buffering is the correct term since it's what speedrunner uses when they really need a frame perfect trick to be exucuted

garrettjoecox commented:
This is only part of the "pause buffering" trick though, at least per https://www.zeldaspeedruns.com/oot/tech/pause-buffering

This helps them easily get to the frame they need, but if I'm understanding correctly actual "pause buffering" is timing your inputs correctly so that they are executed on the next frame after unpausing. Which this cheat does not make any easier aside from getting them up to the frame they need.

garrettjoecox committed:
Add pause input buffering

garrettjoecox committed:
Merge branch 'develop-zhora' of https://github.com/HarbourMasters/Shipwright into cheat-easy-pause-buffer

garrettjoecox commented:
I adjusted the name of the CVars and also added what I believe to be real pause input buffering along side the frame advancing, I'll try and get a few people familiar with frame perfect tricks to verify if this works as expected

Cosmicabuse commented:
I'm trying to test this out, maybe I'm missing the setting to enable it ; I just enabled the cheat. 
I don't think there's any additional unpause lag to be able to actually buffer inputs? Holding L with the cheat = I can't buffer any input in that unpause window, which I want to call frame advance right now.
Pause buffering is being able to buffer any input during the unpause lag easily, on top of another pause if needed. The best way imo would be to have a setting like the Input Lag PR, where you can set in milliseconds what the unpause lag is. 

There's also a quirk to pause buffering you might not know, which is that Stick inputs do NOT depend on that unpause lag buffering window; you can : 
- Pause, hold a control stick direction, 
- Unpause, Pause again after 1 frame

Link will have moved in that control stick direction. Compared to regular button pause buffering, there's a difference : 
- Pause, hold A
- Unpause, nothing happens.

Actual pause buffering allows you to hold A during the unpause lag, so A gets pressed as soon as possible.

garrettjoecox committed:
Fix pause input buffering

garrettjoecox committed:
Merge branch 'develop-zhora' of https://github.com/HarbourMasters/Shipwright into cheat-easy-pause-buffer

garrettjoecox committed:
Changes from feedback

garrettjoecox committed:
Merge branch 'develop-zhora' of https://github.com/HarbourMasters/Shipwright into cheat-easy-pause-buffer

garrettjoecox committed:
Convert frame to bool

garrettjoecox committed:
Handle when pad is nullptr

garrettjoecox committed:
Merge branch 'develop' of https://github.com/HarbourMasters/Shipwright into cheat-easy-pause-buffer

garrettjoecox committed:
Add newline

garrettjoecox committed:
Merge branch 'develop' into cheat-easy-pause-buffer

garrettjoecox commented:
Resolved conflicts

garrettjoecox committed:
Split input buffering into it's own CVAR

garrettjoecox committed:
Split SetShouldBlockGameInput up into to methods
